<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
  <title>Hawkeye Contig View: AMOS Assembly Visual Analytics Tool</title>
</head>
<body>

<a name="top"></a>

<table width="100%" border=0>
<tr>
<td valign=top>
<a href="index.html"><img src="Hawkeye.jpg" width="250" height="160"/></a><br>

<h3>Table of Contents</h3>
<a href="index.html">Abstract</a><br>
<a href="index.html#build">Build &amp; Installation</a><br>
<a href="index.html#launching">Launching Hawkeye</a><br>
<a href="index.html#options">Command Line Options</a><br>
<a href="index.html#sample">Sample Assembly</a><br>
<a href="index.html#contactinformation">Contact Information</a><br>
<a href="index.html#acknowledgements">Acknowledgements</a><br>


<h3>Displays</h3>

<a href="launchpad.html">Launch Pad</a><br>
<table>
<tr><td width="1">&nbsp;</td><td>
<li><a href="launchpad.html#lp-statistics">Statistics</a></li>
<li><a href="launchpad.html#lp-features">Features</a></li>
<li><a href="launchpad.html#lp-libraries">Libraries</a></li>
<li><a href="launchpad.html#lp-scaffolds">Scaffolds</a></li>
<li><a href="launchpad.html#lp-contigs">Contigs</a></li>
<li><a href="launchpad.html#lp-reads">Reads</a></li>
</td></tr></table>

<br>

<a href="scaffold.html">Scaffold View</a><br>
<table>
<tr><td width="1">&nbsp;</td><td>
<li><a href="scaffold.html#sc-misassembly">Mis-assembly Detection</a></li>
<li><a href="scaffold.html#sc-main">Main Display</a></li>
<li><a href="scaffold.html#sc-overview">Overview</a></li>
<li><a href="scaffold.html#sc-controlpanel">Control Panel</a></li>
</td></tr></table>

<br>

<a href="contig.html">Contig View</a><br>
<table>
<tr><td width="1">&nbsp;</td><td>
<li><a href="contig.html#barcode">SNP Barcode</a></li>
<li><a href="contig.html#chromatogram">Raw Chromatograms</a></li>
<li><a href="contig.html#options">Options</a></li>
</td></tr></table>

<br><br>
<br><br>

<p align="left">
<img src="http://sourceforge.net/sflogo.php?group_id=134326&type=5"> <br>
</p>


</td>

<td width="1">&nbsp;</td>


<td>
<h1><a name="contig"></a>Contig View</h1>
<p>
The contig window of the viewer displays the mulitple
alignment of reads within contigs, and lets one view the bases of the
reads
and the consensus sequence. The chromatogram signal, and quality values
of the reads can optionally be displayed, as can the trimmed
unassembled portion of the read. One can quickly and easily navigate to
any position in any contig, or scan contigs for regions of disagreement
between the reads. Alternatively, the consensus sequence of a contig
can be searched by regular expression.
</p>

<p align=center><img alt="correlated" src="correlated.jpg" style="width: 622px; height: 480px;"><br>
The Contig View using the SNP coloring features</p>

<p>
Immediately below the toolbars is the consensus of the contig. A solid
circle above the consensus flags the
position as having a discrepancy in the tiling. Clicking in the
consensus will color and sort the reads based on the base at that
position. In the screenshot above, I clicked on the C at position
11386. The first 2 reads (GDEI048TF and GDEIN20TF) are colored green
because they have an A at that position, whereas the other reads have a
C and are therefore colored blue. This allows one to easily see that
the reads disagree with the other reads in a consensistent way.
Single-clicking on a read displays the chromatogram signal for the
read. The chromatogram displayed in the main window will be stretched
so that the peak occurs aligned with the base calls. Double-clicking on
a read will open a new window with the raw chromatogram (see below). <br>
</p>

<p>
Below the tiling is some summary information, including the filename of
the bank, how many contigs are contained in the bank,
the current contig id, the length of the consensus of the current
contig and the number of reads in the current contig.
</p>

<a href="#top">Return to top</a><br><hr>

<a name="barcode"></a><h2>SNP Barcode</h2>
<p>
Hawkeye can only display at most 50 basepairs from a single read
at a given time and still be legible. Therefore, the font adjustment
utilizes semantic zooming. After switching to below a threshold
instead of the bases displayed as letters, they are displayed as
abstract colored rectangles matching the chromatogram colors. In
addition, instead of displaying every base in every read, it only
displays those positions that disagree with the consensus. This
creates a barcode like display for reads where their patterns of
SNPs can be examined. This reduces the information load of
viewing the multiple alignment and focuses the attention on the
most interesting postions, those that disagree. Furthermore,
very large regions of upto 1000bp can be inspected at once. The
SNP sorting still works in the Barcode display, so clicking in
the consensus will resort and repartition the reads based on that
position.
</p>

<p align=center>
<a href="barcode-bases.png"><img width="300" height="200" src="barcode-bases.png"></a>
<a href="barcode-abstract.png"><img width="300" height="200" src="barcode-abstract.png"></a><br>
Switching from bases to SNP Barcode views.
</p>

<p align=center>
<a href="barcode-zoom.png"><img width="600" height="400" src="barcode-zoom.png"></a><br>
Using the SNP barcode to inspect a very large region
</p>


<a href="#top">Return to top</a><br><hr>

<h2><a name="chromatogram"></a>Raw Chromatograms</h2>
<p>
Double clicking on a read in the tiling displays the read in a separate
chromatogram window. This requires that the chromotogram is available and 
the chromatogram positions have been loaded into the bank, or the positions
are encoded in the trace file. Below we see the chromatogram display for
read DMGLJ45TR. Unlike in the tiling window, the chromatogram displayed
has not been stretched, instead the base calls are spaced to align
with the peaks. The top number is the consensus position. The next line
is the consensus and the current read's base calls are beneath. A solid
circle above the consensus flags there is a discrepancy in the tiling
at that position (as at position 776/802). Below the read's base calls
are the quality values of the base calls. Below the quality values is
the current sequence position (1-based gapped). Below this the
chromatogram signal is displayed, and the chromatogram position is
displayed at the bottom. The window is tinted so that trimmed bases
appear with a dark background- the C at 785/811 is the last base in the
clear range of this read.
</p>

<p align=center>
<a href="rawchromo.png"><img alt="rawchromo" src="rawchromo.png"
 style="width: 506px; height: 141px;"></a><br>
Raw Chromatogram View
</p>

<a href="#top">Return to top</a><br><hr>



<a name="options"></a><h2>Options</h2>

<h3>Menubar</h3>
The File menu has options for opening a bank, setting chromatogram
paths, and displaying summary windows.<br>
The Options menu toggles various settings:<br>
<ul>
  <li><b>Color Bases</b> - Toggle if the bases should be colored red, blue, yellow and green or just black and white</li>
  <li><b>SNP Coloring</b> - Toggles if clicking in the consensus should color the reads based on the base at that position</li>
  <li><b>Show Full Range</b> - Toggles if the trimmed bases should be displayed or not. They are displayed with a red tint.</li>
  <li><b>Show Positions</b> - Toggles if the consensus position should be written at each position or just at every 10 bp.</li>
  <li><b>Show Indicator</b> - Toggles if the carret indicator should be displayed above the consensus position</li>
  <li><b>Show Quality Values</b> - Toggles if the base quality values should be displayed in the tiling</li>
  <li><b>Low Case Low QV</b> - Toggle if bases with low quality (&lt; 30) should be displayed in lower case letters</li>
  <li><b>Highlight Discrepancies</b> - Toggles if bases which disagree with the consensus should be highlighted in purple</li>
  <li><b>Prefetch Chromatograms</b> - Toggles if chromatograms should be displayed without clicking in the tiling.</li>
</ul>


<h3>Toolbars</h3>
<p>
The position box displays the current position (the leftmost displayed
consensus position), and also accepts input for jumping to arbitrary
positions. The up and down arrows next to the spin box step the view
window in 1 base increments. The left and right arrows next to the
position spin box jump to the previous and next position with a 
discrepancy in the reads relative to the current position indicated 
by the caret above the consensus position line.
</p>

<p>
The A+ and A- increases and decreases the font size, and activates
the semantic zooming into the SNP Barcode. The Find
box allows for Qt regular expressions to be input, and the left and
right arrows will search the consensus sequence for that regular
expression, and highlight it in the consensus.
</p>

<a href="#top">Return to top</a><br><hr>


</body>
</html>
