<html>
<head>
  <title>Hawkeye Scaffold View: AMOS Assembly Visual Analytics Tool</title>
</head>
<body>

<a name="top"></a>

<table width="100%" border=0>
<tr>
<td valign=top>
<a href="index.html"><img src="Hawkeye.jpg" width="250" height="160"/></a><br>

<h3>Table of Contents</h3>
<a href="index.html">Abstract</a><br>
<a href="index.html#build">Build &amp; Installation</a><br>
<a href="index.html#launching">Launching Hawkeye</a><br>
<a href="index.html#options">Command Line Options</a><br>
<a href="index.html#sample">Sample Assembly</a><br>
<a href="index.html#contactinformation">Contact Information</a><br>
<a href="index.html#acknowledgements">Acknowledgements</a><br>


<h3>Displays</h3>

<a href="launchpad.html">Launch Pad</a><br>
<table>
<tr><td width="1">&nbsp;</td><td>
<li><a href="launchpad.html#lp-statistics">Statistics</a></li>
<li><a href="launchpad.html#lp-features">Features</a></li>
<li><a href="launchpad.html#lp-libraries">Libraries</a></li>
<li><a href="launchpad.html#lp-scaffolds">Scaffolds</a></li>
<li><a href="launchpad.html#lp-contigs">Contigs</a></li>
<li><a href="launchpad.html#lp-reads">Reads</a></li>
</td></tr></table>

<br>

<a href="scaffold.html">Scaffold View</a><br>
<table>
<tr><td width="1">&nbsp;</td><td>
<li><a href="scaffold.html#sc-misassembly">Mis-assembly Detection</a></li>
<li><a href="scaffold.html#sc-main">Main Display</a></li>
<li><a href="scaffold.html#sc-overview">Overview</a></li>
<li><a href="scaffold.html#sc-controlpanel">Control Panel</a></li>
</td></tr></table>

<br>

<a href="contig.html">Contig View</a><br>
<table>
<tr><td width="1">&nbsp;</td><td>
<li><a href="contig.html#barcode">SNP Barcode</a></li>
<li><a href="contig.html#chromatogram">Raw Chromatograms</a></li>
<li><a href="contig.html#options">Options</a></li>
</td></tr></table>

<br><br>
<br><br>

<p align="left">
<img src="http://sourceforge.net/sflogo.php?group_id=134326&type=5"> <br>
</p>



</td>

<td width="1">&nbsp;</td>


<td>
<h1><a name="scaffold"></a>Scaffold View</h1>
<p>
The scaffold view of the assembly shows how the contigs and
inserts are placed on the scaffold. It uses the mate-pair relationship 
and library sizes to categories the "happiness" of each insert, meaning 
it displays if the paired reads are correctly oriented and at the 
expected distance apart. The threshold distance for a "happy" 
insert can be adjusted by setting the maximum allowed number 
of standard deviations from the mean an insert can be. Details 
on all objects displayed in the Insert view can be found by clicking 
on any object. The mate for any unhappy insert is highlighted by 
right clicking on the read.
</p>

<p>
Along with inserts and contigs, it also plots the read and insert 
coverage at each position along the scaffold, and a measure of 
overall happiness of the inserts called the CE statistic.
The viewer also highlights the location of arbitrary features
along the scaffold. This functionality is currently used to highlight
clues of mis-assembly, such as regions of the genome where the assembly 
has a high occurence of unhappy insert coverage, or regions of high 
density correlated SNPs. Both such events are strong evidence for 
misassembly, and their combination at a location is nearly conclusive
evidence.
</p>

<p align=center>
<a href="scaffold.png"><img width="600" height="375" src="scaffold.png"></a><br>
Scaffold View in Hawkeye<br>
<small>(Click for full-size image)</small>
</p>

<p>
The view is divided into 3 regions. Along the right 
is the control panel (f in the picture) and details panel (g) 
which allows users to filter and set parameters of the display
and see the details of selected objects. Along the bottom left (e) 
is an overview of the entire scaffold, showing the contig placement 
and features. The rest of the display (a-d), the main display,
shows the placement of the contigs (b), features (c), and 
inserts (d), and the statistical levels (a) of the currently 
selected region of the scaffold. The range slider below the overview 
and the magnifying glass tools allow users to select the region 
of interest. The display is interactive, and the details on 
every object are available on demand by clicking on the
object.
</p>
<br>

<a href="#top">Return to top</a><br><hr>

<a name="sc-misassembly"></a><h3>Mis-assembly Detection</h3>
<p>
The combinations of evidence displayed in the scaffold view makes it 
possible to quickly identify mis-assemblies. Consider the region 
displayed below where happymates have been hidden, and k-mer coverage 
is plotted but otherwise uses default parameters. 
</p>

<p>
Our analysis begins at the cluster of yellow compressed inserts. 
Individually, a single compressed mate is not unusual since 
inserts sample the library distribution, but this is an unusually 
large cluster. Similiarly, the cluster of singleton mates (purple)
below the compression is unusually large. Moving up, the small red features
indicated there are multiple correlated SNPs in that same region, but
this is a haploid organism. Further investigation in the 
<a href="contig.html">Contig View</a> should be used to confirm it is
not chance correlations, but given the low background distribution
of correlations, we can assume this is most likely due to mis-assembly.
The bright white in the read coverage heat map indicates this is 
the highest read coverage in the scaffold. The CE Statistic indicates 
a very strong compression, and at -6 is well below the threshold. 
Finally, the spike in kmer coverage in yellow at the top of the
plot indicates this is a complicated repeat region. Every mis-assembly
characteristic has been met and we can conclude undeniably that 
this is a mis-assembly.
</p>

<p> In contrast, note the repeat (high kmer coverage)
on the left of the plot has only 2 compressed mates, no correlated SNPs,
and even coverage. This demonstrates the difficulty in understanding
mis-assemblies: while nearly every mis-assembly occurs in a repeat, not
every repeat is mis-assembled, and the presense of individual mis-assembly
clues such as individual compressed mates are inconclusive. It is only
the combination of evidence that allows one to prove mis-assembly.
</p>

<p align=center>
<a href="sc-compression.png"><img width="550" height="350" src="sc-compression.png"></a><br>
A Mis-assembled region displayed in the Scaffold View
</p>

<a href="#top">Return to top</a><br><hr>


<a name="sc-main"></a><h2>Main Display</h2>

<p>
The main display (a-d) shows the contigs, inserts, features,
and statistical information for a region of the scaffold. The
scaffold of contigs (b) is represented as rectangles appropriately
spaced and sized. The color of the rectangle indicates if the
contig is oriented forward (blue) or reverse (dark blue) in 
the scaffold. Immediately beneath the contigs are two heatmaps. 
The first in purple highlights regions where the insert coverage 
is exceptionally high or low. Similarily, the green track 
highlights high and low read coverage.
</p>

<h3>Features</h3>
<p>
Beneath the heatmaps, are the the feature tracks. Features are regions
of scaffolds or contigs that have been selected for having 
interesting features. AMOS comes bundled with tools for computing
mis-assembly type features, but arbitrary features of any type
can be loaded from a tabbed deliminate file using 'loadFeatures'.
</p>

<p>
The currently available feature types are as follows:<br>
<table border=1>
<tr><td valign=top><b>SNPs</b><br>Red</td><td valign=top>These regions have a high density of correlated SNPs where multiple reads agree with each other, but disagree with the consensus at multiple locations. This occurs for biological reason such as in assemblies of diploid organisms, but also (and more commonly) because of collapsed repeats in the assembly. These regions are computed by the AMOS tools analyzeSNPs and clusterSNPs.</td></tr>
<tr><td valign=top><b>Unitigs</b><br>Light&nbsp;Blue</td><td valign=top>These regions mark the locations of surrogate unitigs created by Celera Assembler, which are regions the assembler computes to be likely repetitive sequence and are often sources of mis-assembly.</td></tr>
<tr><td valign=top><b>AsmQC</b><br>Dark&nbsp;Blue</td><td valign=top>These regions where marked by the AMOS tool asmQC which evaluates the happiness of matepairs in the assembly. More information is available in the <a href="">asmQC documentation</a>.</td></tr>
<tr><td valign=top><b>Breaks</b><br>Pink</td><td valign=top>These regions indicate where a several
singleton reads break their alignment. This can occur because of mis-assembly
where the singleton reads span the junction between the mis-assembled copies.</td></tr>
</table>
</p>

<h3>Coverage and CE Statistic</h3>
<p>
Above the contigs (a) are two plots for coverage levels (top), and 
the CE statistic (bottom). The purple line in the coverage plot
indicates the insert coverage, and the green line in the coverage plot
indicates the read coverage. Mean values in the current scaffold are
displayed as dashed lines. If loaded with -K, kmer coverage will
be plotted in yellow. See <a href="index.html#options">Command Line Options</a> for more information.
</p>

<p>
Beneath the coverage plots is a plot of the CE statistic at that point 
(green and red lines). The CE statistic computes the level of compression 
or expansion for the inserts spanning a particular position. Values near 
zero indicate no deviation, large negative values (&lt;3) indicate 
statistically unlikely compression, and large positive values 
(&gt;3) indicate statistically unlikely expansion, and thus 
flag both compression and expansion type mis-assemblies.
The value is computed on a per library basis, so there will be as many plots
as libaries represented in the scaffold. See the Control Panel Library menu
for the legend of library colors. A manuscript describing the CE
statistic in more detail is in preparation.
</p>

<h3>Inserts</h3>
<p>
Below the feature tracks are the inserts in the scaffold. If possible, 
the mate-pairs are drawn connected by a thin line, while in all cases,
the thick rectangle indicates the position of the read. By default,
the colored and partioned categorically based on their mate happiness.
There are 7 happiness levels:<br>

<table border=1>
<tr><td valign=top><b>Happy</b><br>Green</td><td valign=top>These inserts are happy in terms of both orientation and distance.</td></tr>
<tr><td valign=top><b>Stretched</b><br>Blue</td><td valign=top>These inserts have the correct orientation, but are further apart than expected based on the library distribution and happiness threshold.</td></tr>
<tr><td valign=top><b>Compressed</b><br>Yellow</td><td valign=top>These inserts have the correct orientation, but are closer than expected based on the library distribution and happiness threshold.</td></tr>
<tr><td valign=top><b>Orientation</b><br>Red</td><td valign=top>These inserts have an invalid orientation, either both pointing in the same direction, or away from each other. (Note the reverse is true for transponson reads, and Hawkeye will accomdate for this).</td></tr>
<tr><td valign=top><b>Linking</b><br>Pink</td><td valign=top>This means a read is present in the scaffold, but its mate is in some other scaffold. The thin line indicates the expected position of the mate.</td></tr>
<tr><td valign=top><b>Singleton</b><br>Purple</td><td valign=top>This means a read is present in the scaffold, but its mate is a singleton and not in any contig or scaffold. The thin line indicates the expected position of the mate.</td></tr>
<tr><td valign=top><b>Unmated</b><br>Grey</td><td valign=top>This means the read has no mate provided for unknown reasons, but usually directed closure reads, 454 sequencing data, or a failed mate.</td></tr>
</table>
</p>

<a href="#top">Return to top</a><br><hr>

<a name="sc-overview"></a><h2>Overview</h2>

<p>
The Overview panel (e) shows the entire scaffold and features.
The background is tinted to highlight the currently visible region
in the main display. It is aligned with the range slider 
beneath for selecting regions to display. Clicking in the 
panel recenters the display on the click point.
</p>

<a href="#top">Return to top</a><br><hr>

<a name="sc-controlpanel"></a><h2>Control Panel</h2>
<p>
The top of the control panel (f) controls the pointer. The
arrow allows a user to select objects to get details in the
details box (g). Right clicking an insert with the arrow 
highlights the mate if it is within the current scaffold,
and control-clicking a read jumps to the mate even if the
mate is in a different scaffold. This allows one to follow
chain of mates between scaffolds. The magnifying glass tools
allow one to zoom in or out of the main display.
</p>

<h3>Queries</h3>
<p>
The queries box allows one to control the main display. The
Search box allows one to find any object by regular expresion
on the name (eid or iid) of the object. The Happy Distance
sets the maximum number of standard deviations from the mean
an insert size may be and still be clasifyed as happy.
</p>

<h3>Features</h3>
<p>
Next is a box for the features. Each predefined feature
type has a checkbox and slider. The checkbox controls
if that feature type should be displayed, and the slider
controls how severe the feature has to be to be displayed.
by default, all features are displayed, but the sliders
can be used to show only regions with extreme insert or 
read coverage, for example. The colors of the feature
sliders act as a legend for their display in the main display.
</p>

<h3>Mate Types</h3>
<p>
Next is a series of toggles for the Mate Types, and controls
if that type should be displayed or not. For example, happy
mates are often the least informative, so they can be hidden.
</p>

<h3>Display Toggles</h3>
<p>
Below the mate types are a series of toggles. They are as follows:<br>

<table border=1>
<tr><td valign=top><b>Coverage&nbsp;Stat</b></td><td valign=top> Toggles the display of the coverage level plot.</td></tr>
<tr><td valign=top><b>CE Statistic</b>   </td><td valign=top> Toggles the display of the CE Statistic value.</td></tr>
<tr><td valign=top><b>Connect Mates</b>  </td><td valign=top> Toggles if mates should be draw connected, or as separate reads.</td></tr>
<tr><td valign=top><b>Partition&nbsp;Types</b></td><td valign=top> Toggles if the inserts should be partitioned according to their happiness, or should be draw as compact as possible.</td></tr>
<tr><td valign=top><b>Tint Partition</b> </td><td valign=top> Toggles if a tinted rectangle should be draw behind the different happiness groups, especially if the mate colors are not according to happiness.</td></tr>
</table>
</p>

<h3>Mate Colors</h3>
<p>
Next to the display toggles is a radio button group controlling
how the mates are colored as follows:<br>
<table border=1>
<tr><td valign=top><p align=center><b>Categorical</b><br><small>(Default)</small></p></td><td valign=top>Mates colored by
according to their insert type happiness level.</td></tr>
<tr><td valign=top><b>Continuous</b></td><td valign=top> Happy, compressed, and expanded mates are colored as a function of their deviation from the mean. Insert sizes near the mean will be colored to blend with the background, while large deviations are drawn increasingly bright. This can be more sensitive than categorical coloring for finding clusters of slightly compress/expanded mates. </td></tr>
<tr><td valign=top><b>Linking</b></td><td valign=top>Linking mates are colored according to the contig they link to. Note: there are only 14 colors used, so some colors may repeat if they are a large number of linked contigs. In general, though, the colors will be locally accurate. Click for details to verify.</td></tr>
<tr><td valign=top><b>Library</b></td><td valign=top>Inserts are colored according to their parent library. The legend of colors is displayed below.</td></tr>
</table>
</p>

<h3>Libraries</h3>
<p>
The final region of the control panel is a legend for the libraries.
Each library is listed by iid along with their mean and standard
deviation. The color code is represented by a sample insert, but the
same colors are also used for the CE statistic plot.
</p>

<p align=center>
<img src="sc-librarymenu.png"><br>
<small>Example Library Legend in the Control Panel</small>
</p>

<a href="#top">Return to top</a><br><hr>

</body>
</html>
